judgels-backends/uriel/uriel-app/src/main/java/judgels/uriel/contest/scoreboard/updater/ContestScoreboardUpdater.java

```diff
@@ -0,0 +1,131 @@
+    private final ContestProblemStore contestProblemStore;
+    private final ContestSubmissionStore contestSubmissionStore;
+    private final ScoreboardProcessorRegistry scoreboardProcessorRegistry;
+
+    private Contest contest;
+
+    public ContestScoreboardUpdater(
+            ObjectMapper objectMapper,
+            ContestScoreboardStore contestScoreboardStore,
+            ContestModuleStore contestModuleStore,
+            ContestContestantStore contestContestantStore,
+            ContestProblemStore contestProblemStore,
+            ContestSubmissionStore contestSubmissionStore,
+            ScoreboardProcessorRegistry scoreboardProcessorRegistry) {
+        this.objectMapper = objectMapper;
+        this.contestScoreboardStore = contestScoreboardStore;
+        this.contestModuleStore = contestModuleStore;
+        this.contestContestantStore = contestContestantStore;
+        this.contestProblemStore = contestProblemStore;
+        this.contestSubmissionStore = contestSubmissionStore;
+        this.scoreboardProcessorRegistry = scoreboardProcessorRegistry;
+    }
+
+    public ContestScoreboardUpdater initJob(Contest newContest) throws CloneNotSupportedException {
+        return ((ContestScoreboardUpdater) this.clone()).setContest(newContest);
+    }
+
+    public ContestScoreboardUpdater setContest(Contest contest) {
+        this.contest = contest;
+        return this;
+    }
+
+    @Override
+    @UnitOfWork
+    public void run() {
+        ContestModulesConfig contestModulesConfig = contestModuleStore.getConfig(contest.getJid(), contest.getStyle());
+
+        List<String> problemJids = contestProblemStore.getProblemJids(contest.getJid());
+        Set<String> contestantJids = contestContestantStore.getApprovedContestantJids(contest.getJid());
+        Map<String, String> problemAliasesMap = contestProblemStore.getProblemAliasesByJids(
+                contest.getJid(),
+                new HashSet<>(problemJids));
+        ScoreboardState scoreboardState = new ScoreboardState.Builder()
+                .problemJids(problemJids)
+                .contestantJids(contestantJids)
+                .problemAliases(problemJids.stream().map(problemAliasesMap::get).collect(Collectors.toList()))
+                .build();
+
+        Map<String, Optional<Instant>> contestantStartTimesMap =
+                contestContestantStore.getApprovedContestantStartTimes(contest.getJid());
+
+        List<Submission> submissions = contestSubmissionStore.getSubmissions(
+                contest.getJid(),
+                Optional.empty(),
+                Optional.empty(),
+                Optional.empty())
+                .getPage();
+        generateAndUpsertScoreboard(
+                scoreboardState,
+                contestModulesConfig,
+                contestantStartTimesMap,
+                submissions,
+                ContestScoreboardType.OFFICIAL);
+
+        if (contestModulesConfig.getFrozenScoreboard().isPresent()) {
```

Comment:
You should also check whether the current time has passed the freeze time (can inject `Clock`)